crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1297,6 +1272,128 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
         }
     }
 
+    /// Remove the local timeline item matching the `event_id` or the
+    /// `transaction_id` of `new_event_timeline_item` if it exists.
+    ///
+    /// Let's also try to deduplicate remote evevents. If `VectorDiff`s are the
+    /// inputs of the `Timeline`, this is not necessary, as they are
+    /// generated by the `EventCache`, which supports its own deduplication
+    /// algorithm.
+    //
+    // Note: this method doesn't take `&mut self` to avoid a borrow checker
+    // conflict with `TimelineEventHandler::add_item`. It's fine.
+    fn deduplicate_local_timeline_item(
+        items: &mut ObservableItemsTransaction<'_>,
+        new_event_timeline_item: &mut EventTimelineItem,
+        event_id: Option<&EventId>,
+        transaction_id: Option<&TransactionId>,
+        metadata: &TimelineMetadata,
+        settings: &TimelineSettings,
+    ) -> Option<Arc<TimelineItem>> {
+        /// Transfer `TimelineDetails` that weren't available on the original
+        /// item and have been fetched separately (only `reply_to` for
+        /// now) from `old_item` to `item`, given two items for an event
+        /// that was re-received.
+        ///
+        /// `old_item` *should* always be a local timeline item usually, but it
+        /// can be a remote timeline item.
+        fn transfer_details(new_item: &mut EventTimelineItem, old_item: &EventTimelineItem) {
+            let TimelineItemContent::Message(msg) = &mut new_item.content else { return };
+            let TimelineItemContent::Message(old_msg) = &old_item.content else { return };
+
+            let Some(in_reply_to) = &mut msg.in_reply_to else { return };
+            let Some(old_in_reply_to) = &old_msg.in_reply_to else { return };
+
+            if matches!(&in_reply_to.event, TimelineDetails::Unavailable) {
+                in_reply_to.event = old_in_reply_to.event.clone();
+            }
+        }
```

Comment:
I really don't find the local function to be a win in terms of understanding / readability here, on the contrary it confuses me because it makes the function much larger than it ought to be. Small functions are nice. Could you consider reverting moving this code, please?

Reply:
I could. And I did it.